dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/business/HostAPIImpl.java

```diff
@@ -121,31 +121,24 @@ public Host findDefaultHost(User user, boolean respectFrontendRoles) throws DotS
     @Override
     @CloseDBIfOpened
     public Host resolveHostName(String serverName, User user, boolean respectFrontendRoles) throws DotDataException, DotSecurityException {
-        Host host;
-        final Host cachedHostByAlias = hostCache.getHostByAlias(serverName);
-        if (UtilMethods.isSet(() -> cachedHostByAlias.getIdentifier())) {
-            if (HostCache.CACHE_404_HOST.equals(cachedHostByAlias.getIdentifier())) {
-                return null;
-            }
-            host = cachedHostByAlias;
-        } else {
-            final User systemUser = APILocator.systemUser();
+        Host host = hostCache.getHostByAlias(serverName);
+        User systemUser = APILocator.systemUser();
+
+        if(host == null){
```

Comment:
It looks to me that this is missing the 404 cache check and might cause a performance regression.  We need to keep the 404 cache in mind in our return here.

Here is the original issue that this was intended to fix.


https://github.com/dotCMS/core/issues/27366

Reply:
There is an issue here but maybe not why you think.

Yep I see that although we do not need the full check in here that hostCache.getHostByAlias(serverName) can now return a dummy cache404Contentlet and with the host == null check it will not fall through passing this dummy contentlet which is not correct.

To note though that we do not specifically need to protect the code within the host==null block this is because the caching code is already within the resolveHostNameWithoutDefault and findDefaultHost handling correctly the lookup to the cache and return immediately for a known missing host like "localhost".   The solution here is to actaully remove the initial check against the cache and always run the code in the host == null,   to add the cache checking around this is double checking the cache and produces the inconsistent behavior that triggered the issue.   We do end up with a separate call to findDefaultHost  for an item like localhost that is not in the db, but this should return the cached default host, and the initial call to resolveHostNameWithoutDefault will return immediately and Optional.empty() if there is a 404 entry in the cache preventing the db hit.     Therefore by just having the contents within the host == null you still get all the benefits of the initial issue that modified resolveHostNameWithoutDefault with the extra caching

Reply:
I think... this is fine though I would be way more comfortable to run `ab` tests locally like in the original issue to insure we don't regress.

Reply:
@dsolistorres can you run some tests of the initial issue using this branch before the merge?

Reply:
Code has been updated,  I don't think there should be any adverse effects that were not already there in the original fix, but maybe we could get @dsolistorres to rerun his original checks as I will not be able to do that myself today.

Reply:
Here are the AB testing results:
* Where the hostname does not exist in dotCMS:
```
$ ab -k -n 10000 -c 20 http://non-existing-site.dotcms.com:8080/testpage
This is ApacheBench, Version 2.3 <$Revision: 1913912 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking non-existing-site.dotcms.com (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        dev-station.local|f308ce8c34
Server Hostname:        non-existing-site.dotcms.com
Server Port:            8080

Document Path:          /testpage
Document Length:        765 bytes

Concurrency Level:      20
Time taken for tests:   2.616 seconds
Complete requests:      10000
Failed requests:        0
Non-2xx responses:      10000
Keep-Alive requests:    9910
Total transferred:      11707480 bytes
HTML transferred:       7650000 bytes
Requests per second:    3823.14 [#/sec] (mean)
Time per request:       5.231 [ms] (mean)
Time per request:       0.262 [ms] (mean, across all concurrent requests)
Transfer rate:          4371.03 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       1
Processing:     1    5   3.8      5     218
Waiting:        0    5   3.8      5     217
Total:          1    5   3.8      5     218

Percentage of the requests served within a certain time (ms)
  50%      5
  66%      6
  75%      7
  80%      7
  90%      8
  95%     10
  98%     12
  99%     15
 100%    218 (longest request)
```

* And for an existing hostname:
```
$ ab -k -n 10000 -c 20 http://test-site.dotcms.com:8080/testpage
This is ApacheBench, Version 2.3 <$Revision: 1913912 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking test-site.dotcms.com (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        dev-station.local|f308ce8c34
Server Hostname:        test-site.dotcms.com
Server Port:            8080

Document Path:          /testpage
Document Length:        27094 bytes

Concurrency Level:      20
Time taken for tests:   4.893 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      277920000 bytes
HTML transferred:       270940000 bytes
Requests per second:    2043.66 [#/sec] (mean)
Time per request:       9.786 [ms] (mean)
Time per request:       0.489 [ms] (mean, across all concurrent requests)
Transfer rate:          55466.29 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.2      0       2
Processing:     2    9  11.2      8     215
Waiting:        2    9  11.0      7     215
Total:          2   10  11.2      8     216

Percentage of the requests served within a certain time (ms)
  50%      8
  66%      9
  75%     11
  80%     12
  90%     15
  95%     19
  98%     28
  99%     39
 100%    216 (longest request)
```

Reply:
@dsolistorres WAIT - the ab test looks like returning Non 200 results. Should not happen

Reply:
shouldn't it though?  Since that's the one where hostname doesn't exist should it be a `404` ?

Reply:
@sfreudenthaler when the hostname doesn't exist the request goes to the default site, so the 404 is returned only if the requested page doesn't exist on the default host (that was the case with my first test). In this case I think we want to test the case where the page does exist on the default host, so we compare more similar cases.

Here is a new test that I did without the 404:
* Case 1: the hostname exists
```
$ ab -k -n 10000 -c 20 http://demo.dotcms.com:8080/testpage
This is ApacheBench, Version 2.3 <$Revision: 1913912 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking demo.dotcms.com (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        dev-station.local|f308ce8c34
Server Hostname:        demo.dotcms.com
Server Port:            8080

Document Path:          /testpage
Document Length:        27106 bytes

Concurrency Level:      20
Time taken for tests:   3.709 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      278040000 bytes
HTML transferred:       271060000 bytes
Requests per second:    2695.86 [#/sec] (mean)
Time per request:       7.419 [ms] (mean)
Time per request:       0.371 [ms] (mean, across all concurrent requests)
Transfer rate:          73198.89 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.2      0       3
Processing:     2    7   4.8      6     149
Waiting:        2    7   4.7      6     148
Total:          2    7   4.8      6     149

Percentage of the requests served within a certain time (ms)
  50%      6
  66%      7
  75%      8
  80%      8
  90%     10
  95%     13
  98%     18
  99%     25
 100%    149 (longest request)
```

* Case 2: the hostname doesn't exist, but the requested page exists on the default host:
```
$ ab -k -n 10000 -c 20 http://127.0.0.1:8080/testpage
This is ApacheBench, Version 2.3 <$Revision: 1913912 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 127.0.0.1 (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        dev-station.local|f308ce8c34
Server Hostname:        127.0.0.1
Server Port:            8080

Document Path:          /testpage
Document Length:        27106 bytes

Concurrency Level:      20
Time taken for tests:   2.026 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      278040000 bytes
HTML transferred:       271060000 bytes
Requests per second:    4936.99 [#/sec] (mean)
Time per request:       4.051 [ms] (mean)
Time per request:       0.203 [ms] (mean, across all concurrent requests)
Transfer rate:          134050.76 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       8
Processing:     1    4   2.5      3      58
Waiting:        1    4   2.4      3      58
Total:          1    4   2.5      3      58

Percentage of the requests served within a certain time (ms)
  50%      3
  66%      4
  75%      4
  80%      5
  90%      6
  95%      8
  98%     11
  99%     13
 100%     58 (longest request)
```

Reply:
looks better!